sycl-fusion/jit-compiler/lib/translation/KernelTranslation.cpp

```diff
@@ -0,0 +1,264 @@
+  }
+  return std::move(Result);
+}
+
+llvm::Expected<std::unique_ptr<llvm::Module>>
+KernelTranslator::loadLLVMKernel(llvm::LLVMContext &LLVMCtx,
+                                 SYCLKernelInfo &Kernel) {
+  auto &BinInfo = Kernel.BinaryInfo;
+  llvm::StringRef RawData(reinterpret_cast<const char *>(BinInfo.BinaryStart),
+                          BinInfo.BinarySize);
+  return llvm::parseBitcodeFile(
+      MemoryBuffer::getMemBuffer(RawData, Kernel.Name,
+                                 /* RequiresNullTermnator*/ false)
+          ->getMemBufferRef(),
+      LLVMCtx);
+}
+
+llvm::Expected<std::unique_ptr<llvm::Module>>
+KernelTranslator::loadSPIRVKernel(llvm::LLVMContext &LLVMCtx,
+                                  SYCLKernelInfo &Kernel) {
+  return SPIRVLLVMTranslator::loadSPIRVKernel(LLVMCtx, Kernel);
+}
+
+llvm::Error KernelTranslator::translateKernel(SYCLKernelInfo &Kernel,
+                                              llvm::Module &Mod,
+                                              JITContext &JITCtx,
+                                              BinaryFormat Format) {
+
+  KernelBinary *KernelBin = nullptr;
+  switch (Format) {
+  case BinaryFormat::SPIRV: {
+    llvm::Expected<KernelBinary *> BinaryOrError =
+        translateToSPIRV(Mod, JITCtx);
+    if (auto Error = BinaryOrError.takeError()) {
+      return Error;
+    }
+    KernelBin = *BinaryOrError;
+    break;
+  }
+  case BinaryFormat::PTX: {
+    llvm::Expected<KernelBinary *> BinaryOrError = translateToPTX(Mod, JITCtx);
+    if (auto Error = BinaryOrError.takeError()) {
+      return Error;
+    }
+    KernelBin = *BinaryOrError;
+    break;
+  }
+  default: {
+    return createStringError(
+        inconvertibleErrorCode(),
+        "Failed to translate kernel to unsupported output format");
+  }
+  }
+
+  // Update the KernelInfo for the fused kernel with the address and size of the
+  // SPIR-V binary resulting from translation.
+  SYCLKernelBinaryInfo &FusedBinaryInfo = Kernel.BinaryInfo;
+  FusedBinaryInfo.Format = Format;
+  // Output SPIR-V should use the same number of address bits as the input
+  // SPIR-V. SPIR-V translation requires all modules to use the same number of
+  // address bits, so it's safe to take the value from the first one.
+  FusedBinaryInfo.AddressBits = Mod.getDataLayout().getPointerSizeInBits();
+  FusedBinaryInfo.BinaryStart = KernelBin->address();
+  FusedBinaryInfo.BinarySize = KernelBin->size();
+  return Error::success();
+}
+
+llvm::Expected<KernelBinary *>
+KernelTranslator::translateToSPIRV(llvm::Module &Mod, JITContext &JITCtx) {
+  return SPIRVLLVMTranslator::translateLLVMtoSPIRV(Mod, JITCtx);
+}
+
+llvm::Expected<KernelBinary *>
+KernelTranslator::translateToPTX(llvm::Module &Mod, JITContext &JITCtx) {
+  // FIXME: Can we limit this to the NVPTX specific target?
+  llvm::InitializeAllTargets();
+  llvm::InitializeAllAsmParsers();
+  llvm::InitializeAllAsmPrinters();
+  llvm::InitializeAllTargetMCs();
```

Comment:
```suggestion
  LLVMInitializeNVPTXTargetInfo();
  LLVMInitializeNVPTXTarget();
  LLVMInitializeNVPTXAsmParsers();
  LLVMInitializeNVPTXAsmPrinters();
  LLVMInitializeNVPTXTargetMCs();
```

But these will only exists if the NVPTX backend is built. Probably good to take that into account.

Reply:
Switched to use these specific ones. 

I also added a check in CMake to only enable this part of the code and fusion for the PTX target in general, if the `NVPTX` is included in `LLVM_TARGETS_TO_BUILD`.